timemachine/fe/dummy.py

```diff
@@ -0,0 +1,631 @@
+                bonds_13 = set()
+                bonds_14 = set()
+                for idxs in mutual_bonds:
+                    if len(idxs) == 2:
+                        bonds_12.add(idxs)
+                for idxs in mutual_bonds:
+                    if len(idxs) == 3:
+                        i, j, k = idxs
+                        if (ordered_tuple((i, j)) in bonds_12) and (ordered_tuple((j, k)) in bonds_12):
+                            bonds_13.add(idxs)
+                for idxs in mutual_bonds:
+                    if len(idxs) == 4:
+                        i, j, k, l = idxs
+                        if (
+                            (ordered_tuple((i, j)) in bonds_12)
+                            and (ordered_tuple((j, k)) in bonds_12)
+                            and (ordered_tuple((k, l)) in bonds_12)
+                            and (ordered_tuple((i, j, k)) in bonds_13)
+                            and (ordered_tuple((j, k, l)) in bonds_13)
+                        ):
+                            bonds_14.add(idxs)
+
+                mutual_bonds = bonds_12.union(bonds_13).union(bonds_14)
+
+            anchor_group_ixns.append(mutual_bonds)
+
+        all_agcs.append(anchor_group_candidates)
+        all_agis.append(anchor_group_ixns)
+
+    return dgs, all_agcs, all_agis
+
+
+def generate_optimal_dg_ag_pairs(mol, core, bond_idxs, strict=False):
+    """
+    Generate optimal (dummy group, anchor group) pairs given a list of bonded terms.
+
+    The dummy groups are generated from dummy atoms defined as non-core atoms. The heuristic
+    used attempts to maximize the number of 1-2 terms that can be left on, then the number
+    of 1-3 terms that can be left on, and finally the number of 1-4 terms that can be left on.
+
+    Parameters
+    ----------
+    mol: Chem.Mol
+        Input molecule
+
+    core: list of int
+        Indices for the core atom
+
+    bond_idxs: list of list of ints
+        Input bond_idxs
+
+    strict: bool
+        Whether or not we require bond terms to be present
+        angle terms, and bond and angle terms to be present
+        for torsion terms.
+
+    Returns
+    -------
+    3-tuple of dummy_groups, best_anchor_groups, best_anchor_group_ixns
+        Best anchor group for each dummy group and its interactions are returned
+
+    """
+    dgs, all_agcs, all_agis = generate_dg_ag_pairs(mol, core, bond_idxs, strict)
+
+    picked_agcs = []
+    picked_agis = []
+
+    for agcs, agis in zip(all_agcs, all_agis):
+
+        counts = []
+        for mutual_bonds in agis:
+            bond_count = np.sum([len(idxs) == 2 for idxs in mutual_bonds])
+            angle_count = np.sum([len(idxs) == 3 for idxs in mutual_bonds])
+            torsion_count = np.sum([len(idxs) == 4 for idxs in mutual_bonds])
+            counts.append((bond_count, angle_count, torsion_count))
+
+        best_idx = sorted(zip(range(len(counts)), counts), reverse=True, key=lambda x: x[1])[0][0]
```

Comment:
nit: could use `min` here?
```suggestion
        best_idx = min(zip(range(len(counts)), counts), reverse=True, key=lambda x: x[1])[0]
```

Reply:
I think this should be `max` since `min` doesn't support `reverse=True` - will replace though.